### ISSUE TITLE
Implement comprehensive SciRegressor testing for XGBoost, LightGBM, and CatBoost models

### ISSUE THREAD

**Issue by sandrohuni** [hydrosolutions/long-term-forecasting#10]
## Overview

Implement extensive testing for the SciRegressor models (XGBoost, LightGBM, CatBoost) with comprehensive coverage of hyperparameter tuning, different preprocessing steps, calibration, hindcast, and operational prediction workflows.

## Objective

Create a robust test suite that validates the complete machine learning pipeline for monthly discharge forecasting, ensuring each model works correctly with different preprocessing configurations and follows the logical workflow order.

### Models to Test
- **XGBoost** (`xgb`) - Gradient boosting trees
- **LightGBM** (`lgbm`) - Light gradient boosting machine
- **CatBoost** (`catboost`) - Categorical boosting

### Preprocessing Steps to Test
1. **No Normalization** - Raw data without scaling
2. **Global Normalization** - Standard scaling across all basins
3. **Per-Basin Normalization** - Basin-specific scaling
4. **Long-term Mean Scaling** - Seasonal normalization based on historical means

### Workflow Components to Test
1. **Hyperparameter Tuning** - Optuna-based optimization
2. **Calibration** - Leave-One-Year-Out cross-validation
3. **Hindcast** - Historical validation
4. **Operational Prediction** - Real-time forecasting

#### 1.1 Hyperparameter Tuning Tests
Create individual test functions for each model:
```python
def test_xgb_hyperparameter_tuning_no_normalization()
def test_xgb_hyperparameter_tuning_global_normalization()
def test_xgb_hyperparameter_tuning_per_basin_normalization()
def test_xgb_hyperparameter_tuning_long_term_mean_scaling()

def test_lgbm_hyperparameter_tuning_no_normalization()
def test_lgbm_hyperparameter_tuning_global_normalization()
def test_lgbm_hyperparameter_tuning_per_basin_normalization()
def test_lgbm_hyperparameter_tuning_long_term_mean_scaling()

def test_catboost_hyperparameter_tuning_no_normalization()
def test_catboost_hyperparameter_tuning_global_normalization()
def test_catboost_hyperparameter_tuning_per_basin_normalization()
def test_catboost_hyperparameter_tuning_long_term_mean_scaling()
```

#### 1.2 Calibration Tests
Test calibration functionality with different preprocessing:
```python
def test_xgb_calibration_no_normalization()
def test_xgb_calibration_global_normalization()
def test_xgb_calibration_per_basin_normalization()
def test_xgb_calibration_long_term_mean_scaling()

def test_lgbm_calibration_no_normalization()
def test_lgbm_calibration_global_normalization()
def test_lgbm_calibration_per_basin_normalization()
def test_lgbm_calibration_long_term_mean_scaling()

def test_catboost_calibration_no_normalization()
def test_catboost_calibration_global_normalization()
def test_catboost_calibration_per_basin_normalization()
def test_catboost_calibration_long_term_mean_scaling()
```

#### 1.3 Hindcast Tests
Test historical validation with different preprocessing:
```python
def test_xgb_hindcast_no_normalization()
def test_xgb_hindcast_global_normalization()
def test_xgb_hindcast_per_basin_normalization()
def test_xgb_hindcast_long_term_mean_scaling()

def test_lgbm_hindcast_no_normalization()
def test_lgbm_hindcast_global_normalization()
def test_lgbm_hindcast_per_basin_normalization()
def test_lgbm_hindcast_long_term_mean_scaling()

def test_catboost_hindcast_no_normalization()
def test_catboost_hindcast_global_normalization()
def test_catboost_hindcast_per_basin_normalization()
def test_catboost_hindcast_long_term_mean_scaling()
```

#### 1.4 Operational Prediction Tests
Test real-time forecasting with different preprocessing:
```python
def test_xgb_operational_prediction_no_normalization()
def test_xgb_operational_prediction_global_normalization()
def test_xgb_operational_prediction_per_basin_normalization()
def test_xgb_operational_prediction_long_term_mean_scaling()

def test_lgbm_operational_prediction_no_normalization()
def test_lgbm_operational_prediction_global_normalization()
def test_lgbm_operational_prediction_per_basin_normalization()
def test_lgbm_operational_prediction_long_term_mean_scaling()

def test_catboost_operational_prediction_no_normalization()
def test_catboost_operational_prediction_global_normalization()
def test_catboost_operational_prediction_per_basin_normalization()
def test_catboost_operational_prediction_long_term_mean_scaling()
```

#### 2.1 Complete Workflow Tests
Test the full logical workflow: hyperparameter tuning → calibration → operational forecasting:
```python
def test_xgb_complete_workflow_no_normalization()
def test_xgb_complete_workflow_global_normalization()
def test_xgb_complete_workflow_per_basin_normalization()
def test_xgb_complete_workflow_long_term_mean_scaling()

def test_lgbm_complete_workflow_no_normalization()
def test_lgbm_complete_workflow_global_normalization()
def test_lgbm_complete_workflow_per_basin_normalization()
def test_lgbm_complete_workflow_long_term_mean_scaling()

def test_catboost_complete_workflow_no_normalization()
def test_catboost_complete_workflow_global_normalization()
def test_catboost_complete_workflow_per_basin_normalization()
def test_catboost_complete_workflow_long_term_mean_scaling()
```

#### 2.2 Cross-Model Integration Tests
Test model ensemble and comparison:
```python
def test_multi_model_ensemble_no_normalization()
def test_multi_model_ensemble_global_normalization()
def test_multi_model_ensemble_per_basin_normalization()
def test_multi_model_ensemble_long_term_mean_scaling()
```

### Test Data
- Use synthetic hydro-meteorological data generation
- Ensure reproducible results with fixed random seeds
- Test with multiple basins and time periods

### Validation Criteria
- **Model Training**: Verify models train without errors
- **Hyperparameter Optimization**: Check Optuna trials complete successfully
- **Calibration**: Validate Leave-One-Year-Out cross-validation
- **Predictions**: Ensure predictions are in valid range and format
- **Persistence**: Test model and artifact save/load functionality
- **Performance**: Verify R² scores and other metrics are reasonable

### Error Handling
- Test graceful handling of missing data
- Validate error messages for invalid configurations
- Test recovery from failed hyperparameter optimization

### Test Organization
- Extend existing `tests/test_sciregressor.py`
- Use parametrized tests where appropriate
- Group tests by model type and preprocessing method
- Include comprehensive docstrings

### Mock Strategy
- Mock external dependencies (file I/O, Optuna)
- Use synthetic data for consistent testing
- Mock time-dependent operations for reproducibility

### Configuration Management
- Create test-specific configuration files
- Use temporary directories for test artifacts
- Ensure clean test environment isolation

### Coverage Goals
- **Model Coverage**: 100% coverage for XGBoost, LightGBM, CatBoost
- **Preprocessing Coverage**: All four normalization methods tested
- **Workflow Coverage**: Complete pipeline from tuning to prediction
- **Error Coverage**: Common failure scenarios handled

### Performance Benchmarks
- All tests complete within reasonable time limits
- No memory leaks or resource exhaustion
- Consistent results across multiple runs

### Documentation
- Update `tests/README_SCIREGRESSOR_TESTS.md` with new test descriptions
- Include examples of running specific test subsets
- Document expected test data requirements

### Test Files
- `tests/test_sciregressor.py` - Add comprehensive test functions
- `tests/test_data/` - Add any required test data files
- `tests/configs/` - Add test-specific configuration files

### Documentation
- `tests/README_SCIREGRESSOR_TESTS.md` - Update with new test descriptions
- Add inline docstrings to all test functions

### CI/CD
- Ensure new tests are included in GitHub Actions workflow
- Update test documentation in project README

### Phase 1 (Individual Component Tests)
- **Week 1-2**: Implement hyperparameter tuning tests (12 functions)
- **Week 2-3**: Implement calibration tests (12 functions)
- **Week 3-4**: Implement hindcast tests (12 functions)
- **Week 4-5**: Implement operational prediction tests (12 functions)

### Phase 2 (End-to-End Tests)
- **Week 5-6**: Implement complete workflow tests (12 functions)
- **Week 6-7**: Implement cross-model integration tests (4 functions)
- **Week 7-8**: Testing, debugging, and documentation

## Test Summary

**Total Test Functions**: 64
- **48 individual component tests** (12 per workflow component)
- **12 complete workflow tests** (3 models × 4 preprocessing methods)
- **4 cross-model integration tests** (1 per preprocessing method)

**Models**: XGBoost, LightGBM, CatBoost
**Preprocessing Methods**: No normalization, Global normalization, Per-basin normalization, Long-term mean scaling
**Workflow Components**: Hyperparameter tuning, Calibration, Hindcast, Operational prediction

### Test Data Management
- Use consistent synthetic data across all tests
- Ensure test data represents realistic hydrology scenarios
- Document data generation parameters

### Performance Optimization
- Use appropriate test timeouts
- Consider parallel test execution where safe
- Optimize synthetic data generation for speed

### Maintenance
- Include tests in regular CI/CD pipeline
- Set up automated test result reporting
- Plan for future model additions

This comprehensive testing suite will ensure the SciRegressor models are robust, reliable, and ready for operational deployment in monthly discharge forecasting applications.
